docs_src/design/adr/application/0025-Record-and-Replay.md

```diff
@@ -0,0 +1,128 @@
+# Record and Replay ADR
+
+## Submitters
+- Lenny Goodell (Intel)
+
+## Change Log
+- [Pending](https://github.com/edgexfoundry/edgex-docs/pull/863) (2022-09-20)
+
+## Referenced Use Case(s)
+- [Record and Replay UCR](https://docs.edgexfoundry.org/2.3/design/ucr/Record-and-Replay/)
+
+## Context
+This design involves creating a new Application Service that is responsible for the requirements in the above referenced UCR. This new Application Service will go above and beyond that of a normal Application Service,  thus it is architecturally significant.
+
+## Proposed Design
+A new Application Service will be created with a RESTful API to handle the Record, Replay, Export and Import capabilities. The service will not create or start a Functions Pipeline on start-up as normally done in Application Services. It will wait until the Record request has been received. Once the recording is complete the Functions Pipeline will be stopped. 
+
+!!! note
+    Application Services do not receive data when the Functions Pipelines are stopped.
+
+### Record Endpoint
+
+This `POST` API will start recording data as specified in the [request Data Transfer Object (DTO)](#record-request-dto) defined below. The request handler will validate the DTO and then create a new Functions Pipeline and [Start the Functions Pipeline](https://docs.edgexfoundry.org/2.3/microservices/application/ApplicationServiceAPI/#makeitrun) to process incoming data. 
+
+The Functions Pipeline will contain the following pipeline functions in the following order
+
+- [Filter functions](https://docs.edgexfoundry.org/2.3/microservices/application/BuiltIn/#filtering) if filtering is needed, configured based on the DTO parameters.
+- [Batching pipeline function](https://docs.edgexfoundry.org/2.3/microservices/application/BuiltIn/#batching) configured based on the DTO parameters. This will be used to control the record duration/count. 
+- New pipeline function written to process the batched data once the batching threshold has been exceeded. This function will simply send the recorded data to an async function for processing. 
+
+The async  function receiving the data will first [stop the Functions Pipeline](https://docs.edgexfoundry.org/2.3/microservices/application/ApplicationServiceAPI/#makeitstop) and then save the data for later replay and/or export. It will also determine the list of unique Device Profile and Device Names from the data and store them along side the recorded data. Since app services can receive Events out of order per their timestamps, the saved Event data must be sorted by the Event timestamps. All data will saved in in-memory storage. 
```

Comment:
> All data will saved in in-memory storage. 

This again makes it harder to justify having this component as a service with a networking API. Maybe I misunderstood, but if all events are buffered in memory, it is better to allow execution only on device to an administrator that runs this process in foreground with knowledge of available system resources.

Reply:
> Maybe I misunderstood, but if all events are buffered in memory, it is better to allow execution only on device to an administrator that runs this process in foreground with knowledge of available system resources.

This is a developer tool, not something that is intended to run on a production system.

Reply:
A developer tool which runs as a service needs to be implemented to function reliably in background. 

> Since app services can receive Events out of order per their timestamps, the saved Event data must be sorted by the Event timestamps. All data will saved in in-memory storage.

Is the data stored in memory to simplify sorting? Why not use a key-value such as [LevelDB](https://github.com/syndtr/goleveldb) which persists on disk and inserts in sorted order?

Reply:
A developer tool which runs as a service needs to be implemented to function reliably in background. 

> Since app services can receive Events out of order per their timestamps, the saved Event data must be sorted by the Event timestamps. All data will saved in in-memory storage.

Is the data stored in memory to simplify sorting? Why not use a key-value such as [LevelDB](https://github.com/syndtr/goleveldb) which persists on disk and inserts in sorted order?

Reply:
> Is the data stored in memory to simplify sorting?

No, it is to keep the initial implementation simple (crawl phase). I am not oppose to a more complex storage as a follow on enhancement once we determine it is justified by usage and feedback.

Reply:
If removing specifying the type of storage appease you here then I am fine with leaving it to an implementation detail.

Reply:
I found it clarifying that the ADR said it stored the capture in memory; otherwise, it looked like the design missed something where it would have needed a means to specify a filename to save it as.

Reply:
If you truly believe in-memory recording is good design and worth becoming part of the EdgeX reference architecture for record and reply, then please leave it.

Reply:
in-memory recording could be an easy starting of this service, and the persistence layer can be extended in the future.
however, according to my experience, if the event frequency is high, for example, 100 auto events with 100ms interval, the database access could be an serious bottleneck, even using the memory db like Redis or [ql(an embedded SQL database engine)](https://godoc.org/github.com/cznic/ql).

We did a big change for device-virtual to improve the performance.
https://github.com/edgexfoundry/device-virtual-go/issues/260